Secure inter-cluster query execution (with initial_user as current query user) [v3]

azat opened the pull request:
Changelog category (leave one):
- New Feature

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Secure inter-cluster query execution (with initial_user as current query user)

Detailed description / Documentation draft:

Implemented:
- Add inter-server cluster secret (to verify the query and all related)
- Use initial_user as current query user (to apply correct `*_for_user` settings)

Usage example:
- set `remote_servers.$CLUSTER.secret` to use initial_user as current query user (if omitted/empty then works as before)

v1: #11391
v2: #11498
Fixes: #6843
Fixes: #9751
Cc: @filimonov
Cc: @vitlibar 

P.S. Not sure that this should be `New Feature`, marked it so to highlight this

<details>

HEAD archive:
- 57eef2b8a1e85d40611950755c2294145b73ef1b # BEGINNING
- f74295c61f6bf1577e49724974da34ed2776fa9a
- dfb70cf7d41de725bbc71c8da86d648cb27f8464 (inter-server user marker)

HEAD pre-merge:
-  c1a61eed4362cfb1745a012a37a96358997356f4

</details>

azat commented:
>Integration tests (asan) — fail: 12, passed: 761, error: 0
>Integration tests (release) — fail: 12, passed: 760, error: 0

test_cluster_copier - does not looks related

filimonov commented on src/Client/Connection.cpp:
```diff
@@ -404,6 +405,49 @@ void Connection::sendQuery(
+            data += query;
+            data += query_id;
+            data += user;
+            data += client_info->initial_user;
```
Since we already get some server info in hello packet, may be server can also send some rand string / salt (ideally it also should be signed with secret) which will be used in hashing to avoid possibility of reusing that hash out of that session.

azat replied:
This can't be done in hello packet, since the connection does not bound to specific cluster.

azat replied:
> ideally it also should be signed with secret

This can't be done (actually it is also possible, by sending it for each cluster, but this is too much)
Everything else, makes sense

azat replied:
Although not sure that it is better then IP address/hostname (todo was left for this for now)

filimonov replied:
Hello - server send random string salt only (no need to be bound to cluster, bounded to connection)

Client then use that salt for creating a signature (i.e. `Sha256(cluster name, query id,..., salt)`). That prevents reusing signature outside of that connection (trafic sniffing attacks etc).

Even safer option is to change that salt for every query but it should be server generated, and require more rounds.

filimonov replied:
Just some more things here - actually that type of 'salt' should be rather called nonce (see https://en.wikipedia.org/wiki/Cryptographic_nonce )

The task we did is called MAC/[HMAC](https://en.wikipedia.org/wiki/HMAC), maybe we can make it a bit more standard way to avoid potential pitfalls (with the security it's better to avoid reinventing things when possible) using standard HMAC-SHA256 (anyway we use openssl, which provide different options for HMAC).

For authentication (with nonce / timestamp / token ) - HMAC is used for example in OAuth: https://tools.ietf.org/html/rfc5849#page-17

azat commented:
@vitlibar can you take a look please?

vitlibar commented on src/Server/TCPHandler.cpp:
```diff
@@ -873,17 +875,71 @@ void TCPHandler::receiveQuery()
+            throw NetException("Hash mismatch", ErrorCodes::UNEXPECTED_PACKET_FROM_CLIENT);
+        /// TODO: change error code?
+
+        query_context->setUserSecure(client_info.initial_user, socket().peerAddress());
```
So you're going to call `setUserSecure()` for each query. I don't think it's unnecessary. I suppose it's probably better to call this function once per connection in `receiveHello()` and just check hash in `receiveQuery()`.

azat replied:
But the initial_user may be different for each query, `ClientInfo` (that contains `initial_user`) is passed to the `Connection::sendQuery` and then on the other side of socket `TCPHandler::receiveQuery` it is received and interpreted.

vitlibar replied:
ok.

vitlibar commented on src/Interpreters/Context.cpp:
```diff
@@ -696,6 +696,41 @@ void Context::setUser(const String & name, const String & password, const Poco::
     setSettings(*access->getDefaultSettings());
 }
 
+void Context::setUserSecure(const String & name, const Poco::Net::SocketAddress & address)
```
Too much of copy-pasted code from `Context::setUser()`. Could you please make some common private function instead?

azat replied:
I though about it, but it is kind of in the middle, and I was afraid of making functions too specific, ok will look deeper

vitlibar commented on src/Interpreters/Context.cpp:
```diff
@@ -696,6 +696,41 @@ void Context::setUser(const String & name, const String & password, const Poco::
     setSettings(*access->getDefaultSettings());
 }
 
+void Context::setUserSecure(const String & name, const Poco::Net::SocketAddress & address)
```
In fact this is "insecure" version of the `Context::setUser()` function. I think the name `setUserWithoutCheckingPassword()` would be clearer.

azat replied:
The though was that `setUserSecure` assumes that you are calling it from *secure* context, but can be mixed up, agree

vitlibar commented on src/Server/TCPHandler.cpp:
```diff
@@ -873,17 +875,71 @@ void TCPHandler::receiveQuery()
+    }
+    else
+    {
+        query_context->setInitialRowPolicy();
```
This call of `setInitialRowPolicy()` was added when `initial_user` and `cluster_user` are not the same. If you pass `initial_user` to your version of `Context::setUser()` you won't have to call `setInitialRowPolicy()` in that case.

azat replied:
Indeed, and it does not called for the case when initial_user is used for the query execution

vitlibar commented on src/Server/TCPHandler.cpp:
```diff
@@ -873,17 +892,75 @@ void TCPHandler::receiveQuery()
+            throw NetException("Hash mismatch", ErrorCodes::UNEXPECTED_PACKET_FROM_CLIENT);
+        /// TODO: change error code?
+
+        query_context->setUserWithoutCheckingPassword(client_info.initial_user, socket().peerAddress());
```
I seems we can skip calling `Context::setUser()` in `TCPHandler::receiveHello()` when the `interserver_secret` is specified.

azat replied:
Can't be done, since Hello can't be expanded

vitlibar commented on src/Server/TCPHandler.cpp:
```diff
@@ -873,17 +892,75 @@ void TCPHandler::receiveQuery()
+    std::string received_hash;
+    if (interserver_secret)
+    {
+        readStringBinary(received_cluster, *in);
```
Maybe it's better to pass the cluster's name with the hello package.

azat replied:
Great idea!

azat replied:
Although it can't be done in the Hello package (since it can't be expanded), requires separate package, say `Cluster` (cluster name, salt)

vitlibar replied:
We may have the extended hello package. I mean if the client wants to send some extra data (cluster's name) or receive some extra data (salt) it will send the extended hello package instead of the standard one. If the server has an older version of course it won't able to manage that extended hello package, but an older version is not able to manage interserver secret anyway.

azat replied:
Ok, added `Cluster` packet type for now (I don't see any pros of adding some super packet that will encapsulate information for lots of places, better to add more packets I would say)

vitlibar replied:
Adding a new packet which will be always sent after the hello packet can affect the performance. Though I don't know if we have to worry about the impact.

vitlibar replied:
I've talked about the problem with Alexey Milovidov. So there are two solutions:
1) send an extra packet just as you're doing. The advantage of the solution is that it's easier to implement.
2) send an extended hello packet first (with another packet type), then if the client receives an exception it checks the error code and if the error code is UNEXPECTED_PACKET_FROM_CLIENT the client sends the standard hello packet. The advantage of this solution is that the number of packets will be increased only if the server and the client have different versions.

azat replied:
>Adding a new packet which will be always sent after the hello packet can affect the performance. Though I don't know if we have to worry about the impact.

Doubt that this will have huge impact, since:
- they will be sent only for distributed queries with interserver secret enabled (http over TLS has a huge overhead too)
- and besides clickhouse-server does not built for running tons of queries per second, so one extra package does not looks like a huge problem to me

vitlibar commented on src/Server/TCPHandler.cpp:
```diff
@@ -873,17 +909,68 @@ void TCPHandler::receiveQuery()
+            throw NetException("Hash mismatch", ErrorCodes::UNEXPECTED_PACKET_FROM_CLIENT);
+        /// TODO: change error code?
+
+        query_context->setUserWithoutCheckingPassword(client_info.initial_user, socket().peerAddress());
```
If the interserver secret is used `TCPHandler::receiveHello()` should not `Context::setUser()`. Otherwise you're performing calculation of access rights and default settings twice (because `receiveHello()` calls `Context::setUser()`, and `receiveQuery()` calls `Context::setUserWithoutCheckingPassword()`), and still using the cluster user's name and password. Maybe `TCPHandler::receiveHello` can store user name and password until the first query, when you know whether the interserver secret is used.

azat replied:
>Maybe TCPHandler::receiveHello() can store user name and password

Some people may consider this as a security breach, but I'm fine with this

azat committed:
Drop policy on start in 01308_row_policy_and_trivial_count_query

azat commented:
>AST fuzzer — Fuzzer failed (134). See the logs

```sql
SELECT arrayJoin([[], []]), * FROM numbers(1) GROUP BY number WITH TOTALS
```

Fix - #13625

azat committed:
Use pytest.raises(match=) in test_allowed_client_hosts

Since pytest.raises() produces excinfo not exception, hence at least
str(excinfo.value) should be used to match the exception message.

azat commented:
>Functional stateless tests (release, DatabaseAtomic) — fail: 1, passed: 2151, skipped: 20

`00633_materialized_view_and_too_many_parts_zookeeper` does not looks related

azat commented on src/Server/TCPHandler.cpp:
```diff
@@ -845,6 +889,15 @@ void TCPHandler::receiveQuery()
+    if (hello_credentials)
+    {
+        connection_context.setUser(hello_credentials.user, hello_credentials.password, socket().peerAddress());
+        query_context->setUser(hello_credentials.user, hello_credentials.password, socket().peerAddress());
```
I don't like this place for now, i.e. calling setUser twice, but this is because of how packages are received right now (loop in the receivePacket), this can be fixed (by parsing Cluster package out of the receivePacket loop, although this will not look nice as-is right now, or by replace Hello package with Hellov2/ExtendedHello as was suggested, but I don't like this idea since it depends on exceptions and it is a little bit harder to test), but is this looks really bad or it is just me?

vitlibar replied:
Double call of `setUser()` does look pretty bad for me. What if we move copying `connection_context` to `query_context` into this function?

I like the idea about Hellov2 more because it allows to keep the protocol easy (well, except the case when the server is older than the client). And I think it's better for the future because we can add to Hellov2 some extra field of a variable length so we won't ever have to implement Hellov3.

azat replied:
>Double call of setUser() looks pretty bad for me. What if we move copying connection_context to query_context into this function?

Hm, not sure that copying the whole context is better, but I'm ok with this (since once LDAP will be integrated setUser may take more time)

>I like the idea about Hellov2 more because it allows to keep the protocol easy (well, except the case when the server is older than the client).

I like it, but I don't like complexity in the protocol, not sure that clean protocol should have this...
Although I don't have strong feeling about this (I guess), so you are voting for extending Hello instead of separate Cluster package (like right now) ?

azat replied:
>Hm, not sure that copying the whole context is better, but I'm ok with this (since once LDAP will be integrated setUser may take more time)

Hm, it also may copy the query_id (not a problem, but AFAFICS the connection context should not have any query_id)

azat commented:
>AST fuzzer — Logical error: 'Bad cast from type DB::ColumnVector<int> to DB::ColumnString'.

Can't reproduce the problem

azat commented:
>Functional stateless tests (debug) — fail: 1, passed: 2159, skipped: 13
>Functional stateless tests (memory) — fail: 1, passed: 2164, skipped: 8
>Functional stateless tests (thread) — fail: 1, passed: 2162, skipped: 10

Unrelated (fixed in #13746)

>Integration tests (release) — fail: 1, passed: 782, error: 0

test_storage_rabbitmq - know flacky test

>AST fuzzer — Fuzzer failed (134). See the logs

Known problem with arrayJoin GROUP BY WITH TOTALS (#13625)

azat commented on src/Client/Connection.cpp:
```diff
@@ -108,6 +111,17 @@ void Connection::connect(const ConnectionTimeouts & timeouts)
+        if (server_revision >= DBMS_MIN_REVISION_WITH_INTERSERVER_SECRET && !cluster_secret.empty())
+        {
+#if USE_SSL
+            sendReceiveCluster();
```
@vitlibar as you can see that the Cluster package will be sent only if the inter-server secret had been enabled, and if it was ExtendedHello package, you cannot fallback to Hello anyway, since if the inter-server secret had been configured/enabled server cannot ignore this fact

vitlibar replied:
I'm not sure about that. On the one hand the server can't ignore the interserver secret but on the other hand what if the server is just older than the client? It would be kind of compatibility breaking change.

1. Interserver secret is disabled => we send Hello request => server sends Hello response

2. Interserver secret is enabled => we send ExtendedHello request => server is new => server sends ExtendedHello response

3a. Interserver secret is enabled => we send ExtendedHello request => server is old => server sends UNEXPECTED_PACKET_FROM_CLIENT => we send Hello request => server sends Hello response

-- OR --

3b. Interserver secret is enabled => we send ExtendedHello request => server is old => server sends UNEXPECTED_PACKET_FROM_CLIENT => we stop because we can't let the server ignore the interserver secret.

azat replied:
>I'm not sure about that. On the one hand the server can't ignore the interserver secret but on the other hand what if the server is just older than the client? It would be kind of compatibility breaking change.

Why? One should not enable it if it has clickhouse-server in the cluster w/o this support.

My initial point was, that since you can't ignore the error (in my opinion), there is no much sense in some generic ExtendedHello for this case

azat commented on src/Core/Defines.h:
```diff
@@ -67,6 +67,10 @@
 #define DBMS_MIN_REVISION_WITH_SETTINGS_SERIALIZED_AS_STRINGS 54429
 
+/// Mininum revision supporting interserver secret.
+/// TODO: sync with cmake/autogenerated_versions.txt before merging
```
Just not to forget adjust before merge (right now it matches)

azat replied:
And also `DBMS_TCP_PROTOCOL_VERSION`

azat commented:
>Performance — Timeout :(

@akuzm this PR changes the protocol but does not changes the [revision](https://github.com/ClickHouse/ClickHouse/blob/fc7fb3b10a4a6ae83c30b896f63ba322d233c6c7/cmake/autogenerated_versions.txt#L2) (since it should auto generated as stated by the file), but AFAICS [performance tests uses clickhouse-client from this PR](https://github.com/ClickHouse/ClickHouse/blob/fc7fb3b10a4a6ae83c30b896f63ba322d233c6c7/docker/test/performance-comparison/entrypoint.sh#L123-L124) (and [later](https://github.com/ClickHouse/ClickHouse/blob/fc7fb3b10a4a6ae83c30b896f63ba322d233c6c7/docker/test/performance-comparison/compare.sh#L30)) for both servers, left and right, but it has the same revision and will not work with the server from the "left"

So should I increment revision by myself or performance scripts should be adjusted to use clickhouse-client from the left/right?

akuzm commented:
> > 
> 
> @akuzm this PR changes the protocol but does not changes the [revision](https://github.com/ClickHouse/ClickHouse/blob/fc7fb3b10a4a6ae83c30b896f63ba322d233c6c7/cmake/autogenerated_versions.txt#L2) (since it should auto generated as stated by the file), but AFAICS [performance tests uses clickhouse-client from this PR](https://github.com/ClickHouse/ClickHouse/blob/fc7fb3b10a4a6ae83c30b896f63ba322d233c6c7/docker/test/performance-comparison/entrypoint.sh#L123-L124) (and [later](https://github.com/ClickHouse/ClickHouse/blob/fc7fb3b10a4a6ae83c30b896f63ba322d233c6c7/docker/test/performance-comparison/compare.sh#L30)) for both servers, left and right, but it has the same revision and will not work with the server from the "left"

"right" is the new one, I really should have named them "old" and "new" :) Does it mean that the new client doesn't work with the old server, and the test times out? Looks like a bug, the client from master must work normally with older versions of server. The timeout and absence of error message is a bug in itself.

akuzm commented:
```
2020-08-19 00:02:42	 + clickhouse-client --port 9001 --query 'select 1'
2020-08-19 00:02:42	 Received exception from server (version 20.8.1):
2020-08-19 00:02:42	 Code: 62. DB::Exception: Received from localhost:9001. DB::Exception: Syntax error: failed at position 1 (''): . Unrecognized token. 
2020-08-19 00:02:42	 + kill -0 85
2020-08-19 00:02:42	 + echo .
2020-08-19 00:02:42	 .
```

It just goes into an endless loop, I'll fix the test script.

azat commented:
>"right" is the new one, I really should have named them "old" and "new" :)

+1

> Does it mean that the new client doesn't work with the old server, and the test times out?

Yes (although I'm not sure about timeout)

> Looks like a bug, the client from master must work normally with older versions of server. The timeout and absence of error message is a bug in itself.

The problem is that client and server from the left and right has the same revision, so there is no way to make the compatible, client with previous revision can communicate with the new server, so as new client can communicate with and old server

akuzm commented:
> The problem is that client and server from the left and right has the same revision, so there is no way to make the compatible, client with previous revision can communicate with the new server, so as new client can communicate with and old server.
> That said that no stable release will be affected of course (but I can increment revision manually, although looks like it is done only automatically)

Let's bump the revision manually then, I think it is a correct thing to do when the protocol is changed.

vitlibar commented on src/Server/TCPHandler.cpp:
```diff
@@ -845,6 +889,15 @@ void TCPHandler::receiveQuery()
     state.is_empty = false;
     readStringBinary(state.query_id, *in);
 
+    if (hello_credentials)
```
If the interserver secret is enabled then why do you want to transfer the password? The server shouldn't use the password, it should use the interserver secret for authentication. So you're sending the password while the server doesn't need it.

And if the interserver secret is disabled you're making the things worse because now you're going to keep the password plaintext in memory for a while.

vitlibar replied:
I think it could be better to move the authentication check back into `TCPHandler::receiveHello()` to work as usual if the interserver secret is disabled. And if the interserver secret is enabled the client could send some reserved string like "INTERSERVER SECRET IS ENABLED" instead of a user name and an empty string instead of a password inside the standard hello packet. And if the server is new enough it will understand that such user name means that the server is going to receive an extra information with cluster name and salt. In this case you won't have to keep the credentials in memory.

azat replied:
>I think it could be better to move the authentication check back into TCPHandler::receiveHello() to work as usual if the interserver secret is disabled. And if the interserver secret is enabled the client could send some reserved string like "INTERSERVER SECRET IS ENABLED" instead of a user name and an empty string instead of a password inside the standard hello packet.

User is used for hashing.

> And if the server is new enough it will understand that such user name means that the server is going to receive an extra information with cluster name and salt. In this case you won't have to keep the credentials in memory.

I went another way, for inter-server secret, I sent empty user (anyway it is not allowed) and the real user goes in the password, and if the user is not empty then it calls setUser just in receiveHello, otherwise it will call setUserWithoutCheckingPassword later while receiving the query.

This has some downsides, if the remote shard does not support this then the error will be about empty user.

P.S. I did not added any checks right now (i.e. to avoid `Cluster` package, if setUser is called), just to show the idea (this PR already contains lots of behavior changes and merge commits, that will be not that easy to rebase)

vitlibar replied:
I meant the reserved name like "INTERSERVER SECRET IS ENABLED" (not just empty name) is good because
1) if the server is old so it doesn't support the interserver secret then you will get a quite descriptive error message anyway like `INTERSERVER SECRET IS ENABLED : authentication error`
2) you can send cluster info data right after the standard hello package (i.e. in the same package), and the server will understand from the reserved user name it should read extra data. And sending extra data in the standard package won't break the compatibility in this case cause if the server is old it will stop with authentication error anyway.

Empty name can mean anything, maybe an error, but a reserved name can have a definitive meaning.

azat replied:
I like this idea, done (although looks a little bit tricky, i.e. basically it does not depends from the revision except for query hash now)

azat commented on src/Server/TCPHandler.cpp:
```diff
@@ -836,6 +851,33 @@ bool TCPHandler::receivePacket()
 
+void TCPHandler::receiveCluster()
+{
+    if (!salt.empty() || !cluster.empty())
+        throw NetException("Cluster packet already received", ErrorCodes::UNEXPECTED_PACKET_FROM_CLIENT);
```
This can go away now

azat commented on src/Client/Connection.cpp:
```diff
@@ -198,6 +219,9 @@ void Connection::receiveHello()
             readVarUInt(server_version_patch, *in);
         else
             server_version_patch = server_revision;
+        if (server_revision < DBMS_MIN_REVISION_WITH_INTERSERVER_SECRET && !cluster_secret.empty())
+            throw Exception(ErrorCodes::UNEXPECTED_PACKET_FROM_SERVER,
+                "Server does not support inter-server secret (revision: {})", server_revision);
```
This can go away

azat commented on src/Server/TCPHandler.cpp:
```diff
@@ -873,17 +915,74 @@ void TCPHandler::receiveQuery()
+
+        if (query_context->getUserName() != client_info.initial_user)
+        {
+            connection_context.setUserWithoutCheckingPassword(client_info.initial_user, socket().peerAddress());
+            query_context->setUserWithoutCheckingPassword(client_info.initial_user, socket().peerAddress());
```
Can this be done unconditionally and always, and then only for query_context (need to think about it)

vitlibar replied:
It seems so. However in the interserver-secret mode it's necessary to do something about `connection_settings`, which were calculated as the default settings of the user transferred in the hello package. Probably we can just use query settings instead.

azat committed:
Not only send the exception to the client but also log it

Otherwise there will be only tricky "Unknown packet" exception after
exception during handling INSERT queries.

azat committed:
Break the query loop on UNEXPECTED_PACKET_FROM_CLIENT

azat commented:
@vitlibar can you take a look please? (I'm planning to rebase this once everything will be resolved, but until then it is better to keep all the history)
Btw any chance that this can be included into 20.9?

vitlibar commented:
> @vitlibar can you take a look please? (I'm planning to rebase this once everything will be resolved, but until then it is better to keep all the history)

sorry for the delay

vitlibar commented on src/Client/Connection.cpp:
```diff
@@ -171,8 +174,26 @@ void Connection::sendHello()
+    if (!cluster_secret.empty())
+    {
+        writeStringBinary(INTERSERVER_SECRET_MARKER, *out);
+        writeStringBinary(user /* real user as password, since user is required for hashing */, *out);
```
Can explain a bit more what hashing you mean here?

azat replied:
At first it was just one more variable for hashing, but later I find out that,
there can be the case when intial_user is empty, comment describe this:

```c
/// initial_user can be empty in case of Distributed INSERT via Buffer/Kafka,
/// i.e. when the INSERT is done with the global context (w/o user).
if (!client_info.initial_user.empty())
{
    query_context->setUserWithoutCheckingPassword(client_info.initial_user, socket().peerAddress());
    LOG_DEBUG(log, "User (initial): {}", query_context->getUserName());
}
else
{
    query_context->setUserWithoutCheckingPassword(user_from_hello, socket().peerAddress());
    LOG_DEBUG(log, "User (from hello): {}", query_context->getUserName());
}
```

vitlibar replied:
`StorageBuffer` and `StorageKafka` both use the global context when they want to insert to a target table.
The global context is used for internal operations and it doesn't restraint access rights. That's why the user name is empty in the global context. 

I mean it seems there is no point to call `setUserWithoutCheckingPassword()` for `user_from_hello` here if `client_info.initial_user.empty()`. Because it doesn't add any extra security (you can't even check password). The only thing this call is going to do is bad: it will check that the user defined in the cluster configuration exists on the nodes and it will not allow us to remove that user from the cluster configuration. One of the reasons why I like the idea of the interserver secret is that we will able to get rid of defining the "remote" users in the cluster configuration.

azat replied:
> I mean it seems there is no point to call setUserWithoutCheckingPassword() for user_from_hello here if client_info.initial_user.empty().

But StorageKafka/StorageBuffer can be then passed to the StorageDistributed (via MaterializedView), and it will not have any user, initial_user is empty and user_from_hello will be empty on the remote side, so what user should be used in this case?

vitlibar replied:
No user. If you don't call any kind of `setUser*()` function you'll get a context with no user and no restraints on access rights just like the global context. And it's ok: StorageBuffer doesn't check access rights when it inserts to a target table locally, so it's fine not to check access rights when it works on remote nodes (as long as the interserver secret protects our connections).

azat replied:
Ok

azat replied:
Done

vitlibar commented on src/Client/Connection.cpp:
```diff
@@ -288,6 +309,22 @@ void Connection::forceConnected(const ConnectionTimeouts & timeouts)
+    salt.clear();
+    pcg64_fast rng(randomSeed());
+    UInt64 rand = rng();
+    std::string_view data(reinterpret_cast<char *>(&rand), sizeof(rand));
```
Minor: Maybe the code would be clearer if we had an overload of `encodeSHA256` like this:

```
void encodeSHA256(const void * data, size_t size, unsigned char * out);
```

so we didn't have to use this `std::string_view data`. And a comment here about why you call `salt.resize(32)` would be nice.

azat replied:
> I think the code would be clearer if we had an overload of encodeSHA256 like this:
>void encodeSHA256(const void * data, size_t size, unsigned char * out);

I though about this, but was going to make it separately.
Ok, added (rebasing and keep only final patches and make them clean will be a nightmare)

> And a comment here about why you call salt.resize(32) would be nice.

There are lot's of other places, that don't have it, hence I decided that it is ok (although I was a bit surprised to see magic number 32, but I guess everybody knows that it is sha256):

```c
programs/install/Install.cpp-                std::vector<uint8_t> hash;
programs/install/Install.cpp:                hash.resize(32);
programs/install/Install.cpp-                encodeSHA256(password, hash.data());
--
src/Access/Authentication.h-    Digest hash;
src/Access/Authentication.h:    hash.resize(32);
src/Access/Authentication.h-    ::DB::encodeSHA256(text, hash.data());
--
src/Storages/MergeTree/MergeTreePartition.cpp-    hash.get128(hash_data);
src/Storages/MergeTree/MergeTreePartition.cpp:    result.resize(32);
src/Storages/MergeTree/MergeTreePartition.cpp-    for (size_t i = 0; i < 16; ++i)
```

vitlibar commented on src/Client/Connection.cpp:
```diff
@@ -405,6 +442,40 @@ void Connection::sendQuery(
+            data += query;
+            data += query_id;
+            data += user;
+            data += client_info->initial_user;
```
Why are you sending `user` here? It seems at this point `user` is already sent in the hello package, and it cannot change like `initial_user`.

azat replied:
Just to make hash depends on both users:
- initial
- and from the hello package

To decrease probability of hijacking as much as possible.

vitlibar replied:
User from the hello package is the "remote" user defined in the cluster configuration. If the interserver secret is enabled there is no much sense to keep this users in the cluster configuration. So `user` here will be probably an empty string and so not very useful to prevent hijacking.

azat replied:
Ok, kind of agree.

vitlibar replied:
Further, you could kind of add more security because you can get the "remote" user from the cluster configuration locally instead of transferring it between nodes. However actually it wouldn't add more security because in this case that user name would work exactly as the second interserver secret.

vitlibar commented on src/Interpreters/Context.cpp:
```diff
@@ -670,15 +670,18 @@ void Context::setUser(const String & name, const String & password, const Poco::
     /// This is harmful field that is used only in foreign "Arcadia" build.
-    client_info.current_password = password;
+    if (password)
+        client_info.current_password = password;
```
`client_info.current_password = password.value_or("");`

azat committed:
OpenSSLHelpers cleanup

Add few more specializations for encodeSHA256():
- std::string encodeSHA256(const std::string_view &);
- std::string encodeSHA256(const void *, size_t);
- void encodeSHA256(const void *, size_t, unsigned char *);

vitlibar commented on src/Client/Connection.cpp:
```diff
@@ -288,6 +309,19 @@ void Connection::forceConnected(const ConnectionTimeouts & timeouts)
+
+    writeStringBinary(cluster, *out);
+    writeStringBinary(salt, *out);
+}
```
I think it's useful to add something like version in these fields.
So if in the future we have to change something in the hashing in future we will able to do it easily.

azat replied:
>So if in the future we have to change something in the hashing we will able to do it easily.

But server revision will be enough for this, no?
Adding one more version here does not looks good to me.

vitlibar replied:
ok

vitlibar commented on cmake/autogenerated_versions.txt:
```diff
@@ -1,5 +1,5 @@
 # This strings autochanged from release_lib.sh:
-SET(VERSION_REVISION 54440)
+SET(VERSION_REVISION 54441)
```
It's bumped automatically by our release scripts.

azat replied:
Yes, but due to https://github.com/ClickHouse/ClickHouse/pull/13156#issuecomment-675633803 perf tests will not work correctly, that's why it has been bumped manually

vitlibar replied:
I'm not quite understand what could happen with the performance tests then. There is no changes in the protocol unless the interserver secret is defined. But it's not defined in the performance test. And last time when I changed the protocol I had not to change `cmake/autogenerated_versions.txt` manually.

vitlibar replied:
This change looks unnecessary, can you revert it? And then we'll see if the performance tests will fail this time.

azat replied:
>And last time when I changed the protocol I had not to change cmake/autogenerated_versions.txt manually.

This depends on what changes you are making

>There is no changes in the protocol unless the interserver secret is defined

I'm afraid this is not correct, since now server \*always\* sends hash for the query (if there is no inter-server secret the hash is empty string).

https://github.com/ClickHouse/ClickHouse/pull/13156/files#diff-43ba0f99a5baf9a6a24378246a48bc06R452

But this can be avoided (in the initial version it was not possible, since it does not have indicator of inter-server not in the query package itself)

vitlibar replied:
It's ok because you check `server_revision` before you send the hash (or empty string).

vitlibar replied:
And still bumping `VERSION_REVISION` looks strange. This value is changed by our scripts when we release ClickHouse and there is no point to change it more often than the releases of ClickHouse are created. You change `DBMS_TCP_PROTOCOL_VERSION` and that should be enough.

azat replied:
upstream/master has 54440 (and this PR will have the same w/o this change), but it does not support cluster-secret, and the client will not be able to communicate with the newer server since it expect the hash for each query (and vice-versa)

vitlibar replied:
It's not necessary to support older or newer versions of master, because those versions are not released yet. It's required to make the master compatible only with versions which released or will be released.

vitlibar replied:
I suggest finishing this PR using the current versioning system because other versioning system likely should be a subject of another PR.

vitlibar replied:
No changes to auto-generated version would be more correct (within the current versioning system).

azat replied:
Ok, but note that once this will be merged some PRs may have failed perf tests (the one that do not have this change, but have the same revision)

vitlibar commented on src/Client/Connection.cpp:
```diff
@@ -288,6 +309,19 @@ void Connection::forceConnected(const ConnectionTimeouts & timeouts)
 }
 
+#if USE_SSL
+void Connection::sendCluster()
```
This function also sends salt. Could you rename it into `sendClusterNameAndSalt()` or something.

azat replied:
By Cluster I mean "info for cluster", but I'm ok with naming it sendClusterNameAndSalt

azat replied:
This is pending, I don't want to burn CI resources, so please let me know when you will finish (you can just add summary to your review, i.e. "Request changes" or similar)

azat replied:
Have you catched anything else?

vitlibar commented on src/Server/TCPHandler.cpp:
```diff
@@ -293,6 +295,12 @@ void TCPHandler::runImpl()
+
             /// If a timeout occurred, try to inform client about it and close the session
             if (e.code() == ErrorCodes::SOCKET_TIMEOUT)
                 network_error = true;
```
The INSERT is often followed by data which should be inserted, and those data aren't involved in this hash calculation. Not sure if we should do something about it right now though.

azat replied:
Can you elaborate?

vitlibar replied:
When `node1` sends an INSERT query to `node2` and some bad guy is in the middle of that communication, he can't change the initial user or the query itself, but he can change the data that will be inserted into `node2` because the data are not protected with the hash.

azat replied:
Oh I see, thank you (I though that this was related to that error handling). To protect from this the hash of the INSERTed blocks should be added, but it will heart performance (a lot).

azat replied:
Actually your can and should use ssl to protect from this use case

vitlibar commented on src/Core/Protocol.h:
```diff
@@ -52,6 +52,10 @@ namespace DB
 
+/// Marker of the inter-server secret (passed in the user name)
+/// (anyway user cannot be started with a whitespace)
+const char INTERSERVER_SECRET_MARKER[] = " INTERSERVER SECRET IS ENABLED ";
```
I think we can shorten the name. Just " INTERSERVER SECRET " seems to be enough.

vitlibar commented on src/Core/Defines.h:
```diff
@@ -69,10 +69,10 @@
 
 /// Version of ClickHouse TCP protocol. Set to git tag with latest protocol change.
-#define DBMS_TCP_PROTOCOL_VERSION 54441
+#define DBMS_TCP_PROTOCOL_VERSION 54226
```
That should be assigned to the version with latest protocol change, i.e. 54440.
I know that this value unlikely affects anything, but it's better to change it anyway.

azat committed:
Bump revision manually (to make it compatible with other version from this tag)

As @alexey-milovidov wrote.

P.S. I did not use ./utils/release/release_lib.sh (via ./release), since
it is pretty complex (it fetches tags and so on)

The following macros has been updated:
- VERSION_REVISION -- usually bumped with utils/release/release_lib.sh
  (interesting that even if there were no changes)
- DBMS_MIN_REVISION_WITH_INTERSERVER_SECRET
- DBMS_TCP_PROTOCOL_VERSION -- does not affect anything, some internal
  yandex stuff

vitlibar commented on src/Interpreters/Cluster.cpp:
```diff
@@ -268,18 +283,25 @@ Clusters::Impl Clusters::getContainer() const
 
-    const auto & config_prefix = cluster_name + ".";
+    secret = config.getString(config_prefix + "secret", "");
+    std::erase(config_keys, "secret");
```
Yandex synchronization check says
```
src/Interpreters/Cluster.cpp:299:10: error: no member named 'erase' in namespace 'std'
    std::erase(config_keys, "secret");
    ~~~~~^
1 error generated.
```

Please replace `std::erase` with `boost::range::remove_erase`.

azat replied:
@vitlibar it uses unbundled libstdc++/libc++?

vitlibar replied:
That's quite an unusual build which links ClickHouse with a lot of closed source code and which still can't use most of C++20 features and also.

azat committed:
Cluster: use boost::range::remove_erase() over std::erase (c++20)

@vitlibar:

  "Yandex synchronization check says

  src/Interpreters/Cluster.cpp:299:10: error: no member named 'erase' in namespace 'std'
      std::erase(config_keys, "secret");
      ~~~~~^
  1 error generated.
  Please replace std::erase with boost::range::remove_erase."

  "That's quite an unusual build which links ClickHouse with a lot of closed source code and which still can't use most of C++20 features."

vitlibar commented:
Some performance tests show slow down. Could you help investigate that? Just add an empty commit or rebase/squash your branch to see if there will be the same performance tests which slow down.

azat commented:
>Some performance tests show slow down. Could you help investigate that? Just add an empty commit or rebase/squash your branch to see if there will be the same performance tests which slow down.

Ok, sure.
But if you will look into previous performance report (before merging upstream and std::erase replace) you will not see any of the failures from the last perf test there (that reports contains two queries that are "slower" though) - https://clickhouse-test-reports.s3.yandex.net/13156/dea1890b0ce68068708b30ec1f2f2640af0bd050/performance_comparison/report.html#fail1

P.S. if you are ready to merge, please let me know, I want to make the history cleaner (or at least squash it).

vitlibar commented:
> P.S. if you are ready to merge, please let me know, I want to make the history cleaner (or at least squash it).

I have marked this PR as approved. It does mean I'm ready to merge it :)
But I have to ensure that the PR doesn't break any tests or there is a good explanation why some tests are broken.

azat committed:
Secure inter-cluster query execution (with initial_user as current query user) [v3]

Add inter-server cluster secret, it is used for Distributed queries
inside cluster, you can configure in the configuration file:

  <remote_servers>
      <logs>
          <shard>
              <secret>foobar</secret> <!-- empty -- works as before -->
              ...
          </shard>
      </logs>
  </remote_servers>

And this will allow clickhouse to make sure that the query was not
faked, and was issued from the node that knows the secret. And since
trust appeared it can use initial_user for query execution, this will
apply correct *_for_user (since with inter-server secret enabled, the
query will be executed from the same user on the shards as on initator,
unlike "default" user w/o it).

v2: Change user to the initial_user for Distributed queries if secret match
v3: Add Protocol::Cluster package
v4: Drop Protocol::Cluster and use plain Protocol::Hello + user marker
v5: Do not use user from Hello for cluster-secure (superfluous)

azat commented:
>I have marked this PR as approved. It does mean I'm ready to merge it :)

Ok, great!
rebased history is ready, waiting for perf tests to push.
Or I can submit it via another PR (so that you can diff them), what you will prefer?

vitlibar commented:
> rebased history is ready, waiting for perf tests to push.
You can push it right now, pushing will restart the tests anyway.

> Or I can submit it via another PR (so that you can diff them), what you will prefer?
Let's stay in this PR.

vitlibar commented on src/Interpreters/Context.cpp:
```diff
@@ -686,15 +686,15 @@ void Context::setUser(const String & name, const String & password, const Poco::
     {
         new_access = getAccessControlManager().getContextAccess(*new_user_id, {}, true, settings, current_database, client_info);
-        if (!new_access->isClientHostAllowed() || !new_access->isCorrectPassword(password))
+        if (!new_access->isClientHostAllowed() || (password && !new_access->isCorrectPassword(*password)))
```
Should we check that the client host is allowed if that's a secondary query? It seems we shouldn't.

azat committed:
Do not check client's host when user comes from interserver secret

azat commented:
@akuzm looks like performance tests hanged, can you take a look please?

akuzm commented:
> @akuzm looks like performance tests hanged, can you take a look please?

They are waiting in queue (even now, 16 hour later...). I increased the priority.
